src/libostree/ostree-sysroot.c

```diff
@@ -374,11 +401,75 @@ remount_writable (const char *path, gboolean *did_remount, GError **error)
   return TRUE;
 }
 
+static gboolean
+is_sysroot_invisible (void)
+{
+  struct stat stbuf;
+
+  if (lstat (OTCORE_RUN_OSTREE_PRIVATE "/sysroot-ns", &stbuf) < 0)
+    return FALSE;
+
+  if (lstat ("/sysroot/ostree", &stbuf) == 0)
+    return FALSE;
+
+  return TRUE;
+}
+
+/* Unmount covering tmpfs to make /sysroot visible */
+gboolean
+_ostree_sysroot_ensure_visible (OstreeSysroot *self, GError **error)
+{
+  if (!ostree_sysroot_initialize (self, error))
+    return FALSE;
+
+  /* Do nothing if no mount namespace is in use */
+  if (!self->mount_namespace_in_use)
+    return TRUE;
+
+  /* If we aren't operating on a booted system, then we don't
+   * do anything with mounts.
+   */
+  if (!self->root_is_ostree_booted)
+    return TRUE;
+
+  /* Handle invisible sysroot */
+  if (is_sysroot_invisible ())
+    {
+      glnx_autofd int sysroot_ns_fd = open (OTCORE_RUN_OSTREE_PRIVATE "/sysroot-ns", O_RDONLY);
+      if (sysroot_ns_fd < 0)
+        return FALSE;
+
+      glnx_autofd int cur_ns_fd = open("/proc/self/ns/mnt", O_RDONLY);
+      if (cur_ns_fd < 0)
+        return FALSE;
+
+      if (setns (sysroot_ns_fd, CLONE_NEWNS) < 0)
```

Comment:
One thing that will be a serious problem for several extant callers is one can't `setns` with a multithreaded process. And both rpm-ostree and bootc for example are multithreaded.

To fix this we'll need to do the fork() + socketpair passing FD I mention below I think.

Reply:
Oh, I see your suggestion of doing
> ExecStart=+/usr/bin/ostree admin nsenter --exec /usr/bin/rpm-ostree start-daemon

Would work around this by having the remounting happen external. And this is also why we'd need to mount `/sysroot` into the new namespace.

But...we could also require callers like rpm-ostree and bootc to invoke the shared library API before becoming multithreaded. Unfortunately a fair bit of stuff happens in `ostree_sysroot_new_default() + ostree_sysroot_set_mount_namespace_in_use (true)`...

Reply:
I think after 160bf13 and e0eb4df, the library can work out of box in a multi-threading environment. Mount namespace is a per-thread attribute, which is tricky to handle. In these two commits, I wrote some code to ensure threads that call the library API can all enter private mount namespaces (not the same one, though).

Reply:
An alternative way is to only use sysroot_fd and boot_fd everywhere and not use paths. We can grab the fd using `open_tree` directly and do not need to enter a mount namespace. However, I'm afraid the caller may still use paths somewhere and it will break the compatibility.